ambry-store/src/main/java/com/github/ambry/store/BootstrapController.java

```diff
@@ -0,0 +1,203 @@
+package com.github.ambry.store;
+
+import com.github.ambry.clustermap.ClusterParticipant;
+import com.github.ambry.clustermap.PartitionId;
+import com.github.ambry.clustermap.PartitionStateChangeListener;
+import com.github.ambry.clustermap.ReplicaId;
+import com.github.ambry.clustermap.StateModelListenerType;
+import com.github.ambry.clustermap.StateTransitionException;
+import com.github.ambry.config.StoreConfig;
+import com.github.ambry.server.StoreManager;
+import java.io.IOException;
+import java.util.Map;
+import java.util.regex.Pattern;
+import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.github.ambry.clustermap.StateTransitionException.TransitionErrorCode.*;
+
+
+public class BootstrapController {
+
+  private final String BOOTSTRAP_IN_PROGRESS_FILE_NAME;
+  private final String FILECOPY_IN_PROGRESS_FILE_NAME;
+  private final Pattern allLogSegmentFilesPattern = Pattern.compile("\\d+\\.log");
+  private final StoreConfig storeConfig;
+  private final StoreManager storeManager;
+  private final PartitionStateChangeListener storageManagerListener;
+  private final PartitionStateChangeListener fileCopyManagerListener;
+
+  private static final Logger logger = LoggerFactory.getLogger(BootstrapController.class);
+
+  public BootstrapController(
+      @Nonnull StoreManager storeManager,
+      @Nonnull StoreConfig storeConfig,
+      @Nonnull ClusterParticipant clusterParticipant) {
+    this.storeConfig = storeConfig;
+    this.storeManager = storeManager;
+
+    this.BOOTSTRAP_IN_PROGRESS_FILE_NAME = storeConfig.storeBootstrapInProgressFile;
+    this.FILECOPY_IN_PROGRESS_FILE_NAME = storeConfig.storeFileCopyInProgressFileName;
+
+    clusterParticipant.registerPartitionStateChangeListener(
+        StateModelListenerType.BootstrapControllerListener, new BootstrapControllerImpl());
+
+    Map<StateModelListenerType, PartitionStateChangeListener> partitionStateChangeListeners =
+        storeManager.getPrimaryClusterParticipant().getPartitionStateChangeListeners();
```

Comment:
We should add null handling for storeManager.getPrimaryClusterParticipant()

Reply:
Added another catch block and we now throw `StateTransitionException` with type `BootstrapControllerFailure`.
Technically yes, `getPrimaryClusterParticipant()` can return null ClusterParticipant.